File: from-0501-to-0600/0501-find-mode-in-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func dfs(node *TreeNode, cnt *map[int]int) {
    if node == nil {
            return
        }
        (*cnt)[node.Val] += 1
        dfs(node.Left, cnt)
        dfs(node.Right, cnt)
}

func findMode(root *TreeNode) []int {
    cnt := make(map[int]int)
    
    dfs(root, &cnt)
    mode := -1000001
    for _, val := range cnt {
        if val > mode {
            mode = val
        }
    }
    result := []int{}
    for key, val := range cnt {
        if val == mode {
            result = append(result, key)
        }
    }
    return result
}

File: from-0501-to-0600/0501-find-mode-in-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        c = defaultdict(int)
        def dfs(node):
            if not node:
                return
            nonlocal c
            c[node.val] += 1
            dfs(node.left)
            dfs(node.right)
            
        dfs(root)
        
        counts = sorted(c.items(), key=lambda x: x[1], reverse=True)
        res = []
        mode = counts[0][1]
        p = 0
        while p < len(counts) and counts[p][1] == mode:
            res.append(counts[p][0])
            p += 1
            
        return res
            